File: polyglot/go/timeline_builder.go
package timeline_builder

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
	"time"
)

// Severity represents the severity level of a vulnerability
type Severity string

const (
	SecHigh   Severity = "HIGH"
	SecMedium Severity = "MEDIUM"
	SecLow    Severity = "LOW"
	SecInfo   Severity = "INFO"
)

// VulnerabilityEvent holds a single vulnerability timeline event
type VulnerabilityEvent struct {
	CVEID       string `json:"cve_id,omitempty"`
	Component   string `json:"component"`
	Version     string `json:"version"`
	DiscoveryAt time.Time `json:"discovery_at"`
	ReportedAt  time.Time `json:"reported_at"`
	FixedAt     *time.Time `json:"fixed_at,omitempty"`
	Severity    Severity  `json:"severity"`
	Description string `json:"description,omitempty"`
}

// TimelineResult holds the complete timeline output
type TimelineResult struct {
	Events      []VulnerabilityEvent
	TotalDays   int
	AvgPatchWindow float64
	MaxPatchWindow int
	MinPatchWindow int
	Summary     map[Severity]int
}

// Config holds optional configuration for building timelines
type Config struct {
	Timezone    string // Default: "UTC"
	IncludeNull bool   // Include events with missing dates
}

// EventSource represents where the raw data comes from
type EventSource interface {
	Read() ([]VulnerabilityEvent, error)
}

// JSONSource reads vulnerability events from a JSON file or stream
type JSONSource struct {
	Path string
	Reader io.Reader
}

func (j *JSONSource) Read() ([]VulnerabilityEvent, error) {
	var events []VulnerabilityEvent
	
	if j.Path != "" {
		data, err := os.ReadFile(j.Path)
		if err != nil {
			return nil, fmt.Errorf("reading file: %w", err)
		}
		err = json.Unmarshal(data, &events)
	} else if j.Reader != nil {
		var data []byte
		data, err := io.ReadAll(j.Reader)
		if err != nil {
			return nil, fmt.Errorf("reading stream: %w", err)
		}
		err = json.Unmarshal(data, &events)
	} else {
		return events, nil
	}

	if len(events) == 0 {
		return events, nil
	}

	// Normalize all timestamps to UTC
	for i := range events {
		events[i].DiscoveryAt = events[i].DiscoveryAt.UTC()
		events[i].ReportedAt = events[i].ReportedAt.UTC()
		if events[i].FixedAt != nil {
			events[i].FixedAt = (*events[i].FixedAt).UTC()
		}
	}

	return events, nil
}

// FileSource reads from a file path directly
type FileSource struct {
	Path string
}

func (f *FileSource) Read() ([]VulnerabilityEvent, error) {
	data, err := os.ReadFile(f.Path)
	if err != nil {
		return nil, fmt.Errorf("reading file %q: %w", f.Path, err)
	}
	
	var events []VulnerabilityEvent
	err = json.Unmarshal(data, &events)
	if err != nil {
		return nil, fmt.Errorf("parsing JSON: %w", err)
	}

	for i := range events {
		events[i].DiscoveryAt = events[i].DiscoveryAt.UTC()
		events[i].ReportedAt = events[i].ReportedAt.UTC()
		if events[i].FixedAt != nil {
			events[i].FixedAt = (*events[i].FixedAt).UTC()
		}
	}

	return events, nil
}

// BuildTimeline creates a sorted timeline from raw events
func BuildTimeline(events []VulnerabilityEvent, cfg Config) (TimelineResult, error) {
	if len(events) == 0 {
		return TimelineResult{}, nil
	}

	// Sort by discovery date first, then reported, then fixed
	sort.SliceStable(events, func(i, j int) bool {
		if !events[i].DiscoveryAt.Equal(events[j].DiscoveryAt) {
			return events[i].DiscoveryAt.Before(events[j].DiscoveryAt)
		}
		if !events[i].ReportedAt.Equal(events[j].ReportedAt) {
			return events[i].ReportedAt.Before(events[j].ReportedAt)
		}
		if events[i].FixedAt == nil || events[j].FixedAt == nil {
			return events[i].FixedAt != nil
		}
		return *events[i].FixedAt.Before(*events[j].FixedAt)
	})

	result := TimelineResult{
		Events: events,
		Summary: make(map[Severity]int),
	}

	now := time.Now()
	if cfg.Timezone != "" {
		loc, err := time.LoadLocation(cfg.Timezone)
		if err == nil {
			now = now.In(loc)
		} else {
			now = now.UTC()
		}
	}

	totalDays := 0
	var patchWindows []int64
	minWindow := int64(365 * 24 * 7) // Default to 1 week
	maxWindow := int64(-1)

	for _, e := range events {
		result.Summary[e.Severity]++

		if !e.DiscoveryAt.IsZero() && !e.ReportedAt.IsZero() {
			days := e.ReportedAt.Sub(e.DiscoveryAt).Hours() / 24.0
			totalDays += int(days)
		}

		if e.FixedAt != nil && !e.DiscoveryAt.IsZero() {
			window := e.FixedAt.Sub(e.DiscoveryAt).Hours() / 24.0
			patchWindows = append(patchWindows, int64(window))
			
			if window < minWindow {
				minWindow = int64(window)
			}
			if window > maxWindow {
				maxWindow = int64(window)
			}
		}

		// Calculate total span from first discovery to now or last fixed
		start := e.DiscoveryAt
		end := now
		if !e.FixedAt.IsZero() && end.Before(*e.FixedAt) {
			end = *e.FixedAt
		}
		
		totalDays += int(end.Sub(start).Hours()) / 24.0
	}

	result.TotalDays = totalDays
	
	var sum float64
	count := 0
	for _, w := range patchWindows {
		sum += float64(w)
		count++
	}
	
	if count > 0 {
		result.AvgPatchWindow = sum / float64(count)
		result.MaxPatchWindow = int(maxWindow)
		result.MinPatchWindow = int(minWindow)
	}

	return result, nil
}

// FormatEvent creates a human-readable string for a single event
func FormatEvent(e VulnerabilityEvent) string {
	var parts []string
	
	if !e.DiscoveryAt.IsZero() {
		parts = append(parts, fmt.Sprintf("Discovered: %s", e.DiscoveryAt.Format("2006-01-02 15:04"))[:19])
	}
	
	if !e.ReportedAt.IsZero() {
		parts = append(parts, fmt.Sprintf("Reported:   %s", e.ReportedAt.Format("2006-01-02 15:04"))[:19])
	}
	
	if e.FixedAt != nil && !e.FixedAt.IsZero() {
		parts = append(parts, fmt.Sprintf("Fixed:      %s", (*e.FixedAt).Format("2006-01-02 15:04"))[:19])
	} else if e.Severity != SecInfo {
		parts = append(parts, "Pending fix")
	}

	if len(parts) == 0 {
		return fmt.Sprintf("%s [%s] %s", e.Component, e.Severity, e.Description)
	}

	return strings.Join(parts, ", ") + " | " + e.Component + " v" + e.Version + " (" + string(e.Severity) + ")"
}

// FormatTimelineResult creates a formatted output for the complete timeline
func FormatTimelineResult(result TimelineResult) string {
	var sb strings.Builder
	
	sb.WriteString("=== VULN TIMELINE SUMMARY ===\n\n")
	
	if len(result.Events) == 0 {
		sb.WriteString("No events found.\n")
		return sb.String()
	}

	sb.WriteString(fmt.Sprintf("Total Events:    %d\n", len(result.Events)))
	sb.WriteString(fmt.Sprintf("Time Span:      %.1f days\n", float64(result.TotalDays)/24.0))

	if result.AvgPatchWindow > 0 {
		sb.WriteString(fmt.Sprintf("Avg Patch Window: %.2f days\n", result.AvgPatchWindow))
		sb.WriteString(fmt.Sprintf("Min Patch Window: %d days\n", result.MinPatchWindow))
		sb.WriteString(fmt.Sprintf("Max Patch Window: %d days\n", result.MaxPatchWindow))
	}

	sb.WriteString("\n--- BY SEVERITY ---\n")
	for sev, count := range []Severity{SecHigh, SecMedium, SecLow, SecInfo} {
		if result.Summary[sev] > 0 {
			sb.WriteString(fmt.Sprintf("  %s: %d\n", sev, result.Summary[sev]))
		}
	}

	sb.WriteString("\n--- DETAILED TIMELINE ---\n")
	
	for _, e := range result.Events {
		sb.WriteString(fmt.Sprintf("%-12s %-8s ", 
			e.Component[:min(30, len(e.Component))],
			string(e.Severity)))
		
		if !e.DiscoveryAt.IsZero() && !e.ReportedAt.IsZero() {
			days := e.ReportedAt.Sub(e.DiscoveryAt).Hours() / 24.0
			sb.WriteString(fmt.Sprintf("Rpt: %.1fd", days))
		} else if !e.DiscoveryAt.IsZero() {
			sb.WriteString("Discovered")
		}
		
		if e.FixedAt != nil && !e.FixedAt.IsZero() {
			days := (*e.FixedAt).Sub(e.DiscoveryAt).Hours() / 24.0
			sb.WriteString(fmt.Sprintf(" Fix: %.1fd", days))
		} else if e.Severity == SecInfo {
			sb.WriteString(" Pending")
		}
		
		sb.WriteString("\n")
	}

	return sb.String()
}

// SourceFromPath creates a JSONSource from a file path string
func SourceFromPath(path string) EventSource {
	return &FileSource{Path: path}
}

// BuildAndFormat is the main convenience function for quick usage
func BuildAndFormat(events []VulnerabilityEvent, cfg Config) (string, error) {
	result, err := BuildTimeline(events, cfg)
	if err != nil {
		return "", fmt.Errorf("building timeline: %w", err)
	}
	
	return FormatTimelineResult(result), nil
}

// Demo function for testing and demonstration purposes
func RunDemo() {
	// Sample vulnerability data
	sampleData := []VulnerabilityEvent{
		{
			CVEID:       "CVE-2024-1234",
			Component:   "libssl",
			Version:     "3.2.1",
			DiscoveryAt: time.Date(2024, 1, 15, 10, 0, 0, 0, time.UTC),
			ReportedAt:  time.Date(2024, 1, 28, 14, 30, 0, 0, time.UTC),
			FixedAt:     &time.Time{}, // Will be set below
			Severity:    SecHigh,
			Description: "Buffer overflow in TLS handshake",
		},
		{
			CVEID:       "CVE-2024-5678",
			Component:   "nginx",
			Version:     "1.25.3",
			DiscoveryAt: time.Date(2024, 2, 1, 9, 0, 0, 0, time.UTC),
			ReportedAt:  time.Date(2024, 2, 15, 16, 0, 0, 0, time.UTC),
			FixedAt:     &time.Time{}, // Will be set below
			Severity:    SecMedium,
			Description: "HTTP/2 header injection",
		},
		{
			CVEID:       "CVE-2024-9999",
			Component:   "redis-server",
			Version:     "7.2.0",
			DiscoveryAt: time.Date(2024, 3, 10, 8, 0, 0, 0, time.UTC),
			ReportedAt:  time.Date(2024, 3, 12, 10, 0, 0, 0, time.UTC),
			FixedAt:     &time.Time{}, // Will be set below
			Severity:    SecLow,
			Description: "Minor memory leak in replication",
		},
	}

	// Set fixed dates for demo
	now := time.Now()
	fixedDates := []time.Time{
		time.Date(2024, 1, 31, 12, 0, 0, 0, time.UTC), // libssl fixed in ~45 days
		time.Date(2024, 2, 28, 18, 0, 0, 0, time.UTC), // nginx fixed in ~43 days  
		now,                                          // redis still pending
	}

	for i := range sampleData {
		sampleData[i].FixedAt = &fixedDates[i]
	}

	fmt.Println("=== VULN TIMELINE BUILDER DEMO ===\n")
	
	result, err := BuildTimeline(sampleData, Config{Timezone: "UTC"})
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}

	output, _ := FormatTimelineResult(result)
	fmt.Println(output)
	
	// Calculate and display key metrics
	fmt.Println("\n=== KEY METRICS ===")
	fmt.Printf("  Average Patch Window: %.2f days\n", result.AvgPatchWindow)
	fmt.Printf("  Min Patch Window: %d days\n", result.MinPatchWindow)
	fmt.Printf("  Max Patch Window: %d days\n", result.MaxPatchWindow)
}

func main() {
	// Run the demo when executed directly
	RunDemo()
	
	// Example usage with file input
	if len(os.Args) > 1 {
		cfg := Config{Timezone: "UTC"}
		
		source := SourceFromPath(os.Args[1])
		events, err := source.Read()
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading events: %v\n", err)
			os.Exit(1)
		}

		result, err := BuildTimeline(events, cfg)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error building timeline: %v\n", err)
			os.Exit(1)
		}

		output, _ := FormatTimelineResult(result)
		fmt.Println(output)
		
		// Output JSON summary to stdout for piping
		jsonSummary, _ := json.MarshalIndent(map[string]interface{}{
			"total_events": len(result.Events),
			"time_span_days": result.TotalDays,
			"avg_patch_window": result.AvgPatchWindow,
			"severity_summary": result.Summary,
		}, "", "  ")
		fmt.Printf("\n=== JSON SUMMARY ===\n%s\n", string(jsonSummary))
	}
}

File: polyglot/python/timeline_builder.py
"""
polyglot/python/timeline_builder.py

Vulnerability disclosure timeline builder with patch-window metrics.

Builds structured timelines from CVE data, calculates patch windows,
and outputs reports in multiple formats (text, JSON, HTML).
"""

from __future__ import annotations

import argparse
import csv
import json
import sys
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import date, timedelta, datetime
from enum import Enum
from html import escape
from pathlib import Path
from typing import Any


class SeverityLevel(Enum):
    CRITICAL = 10
    HIGH = 8
    MEDIUM = 5
    LOW = 2
    INFO = 1

    @classmethod
    def from_string(cls, s: str) -> SeverityLevel:
        mapping = {
            "CRITICAL": cls.CRITICAL,
            "HIGH": cls.HIGH,
            "MEDIUM": cls.MEDIUM,
            "LOW": cls.LOW,
            "INFO": cls.INFO,
        }
        return mapping.get(s.upper(), cls.INFO)

    def __str__(self) -> str:
        return self.name


@dataclass(order=True)
class VersionRange:
    """Represents a range of affected versions."""
    
    min_version: str | None = field(default=None, compare=False)
    max_version: str | None = field(default=None, compare=False)
    is_range: bool = field(default=False, compare=False)

    def __str__(self) -> str:
        if self.is_range and self.min_version and self.max_version:
            return f"{self.min_version} - {self.max_version}"
        return self.min_version or "unknown"


@dataclass
class Vulnerability:
    """A single vulnerability record."""

    cve_id: str
    title: str
    discovered_date: date | None = None
    patched_date: date | None = None
    affected_versions: list[VersionRange] = field(default_factory=list)
    severity: SeverityLevel = SeverityLevel.INFO
    status: str = "open"  # open, patch, retired

    @property
    def days_since_discovery(self) -> int | None:
        if self.patched_date and self.discovered_date:
            return (self.patched_date - self.discovered_date).days
        return None

    @property
    def is_patch_available(self) -> bool:
        return self.patched_date is not None

    @property
    def patch_window_days(self) -> int | None:
        if self.is_patch_available and self.discovered_date:
            return (self.patched_date - self.discovered_date).days
        return None

    @property
    def severity_score(self) -> float:
        scores = {s.name: s.value for s in SeverityLevel}
        return scores.get(self.severity.name, 1.0)


@dataclass
class TimelineEntry:
    """A single entry in the timeline output."""

    date: date
    event_type: str  # "discovered", "patched", "retired"
    cve_id: str | None = None
    title: str | None = None
    severity: SeverityLevel = SeverityLevel.INFO


class TimelineBuilder:
    """Builds vulnerability disclosure timelines from raw data."""

    def __init__(self, config: dict[str, Any] | None = None):
        self.config = config or {}
        self.entries: list[TimelineEntry] = []
        self.vulnerabilities: list[Vulnerability] = []

    @classmethod
    def from_csv(cls, filepath: str | Path) -> TimelineBuilder:
        """Load vulnerabilities from a CSV file with columns:
        
        cve_id,title,discovered_date,patched_date,affected_versions,severity,status
        
        Dates can be YYYY-MM-DD or epoch. Versions are comma-separated like 1.0-2.5,3.0+.
        Severity is one of CRITICAL,HIGH,MEDIUM,LOW,INFO.
        """
        builder = cls()

        with open(filepath, newline="", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            
            for row in reader:
                vuln = cls._parse_row(row)
                if vuln:
                    builder.vulnerabilities.append(vuln)

        return builder

    @classmethod
    def from_dicts(cls, data: list[dict]) -> TimelineBuilder:
        """Build timeline from a list of dictionaries."""
        builder = cls()
        
        for row in data:
            vuln = cls._parse_row(row)
            if vuln:
                builder.vulnerabilities.append(vuln)

        return builder

    @staticmethod
    def _parse_row(row: dict[str, Any]) -> Vulnerability | None:
        """Parse a single row into a Vulnerability object."""
        
        cve_id = str(row.get("cve_id", "")).strip()
        if not cve_id:
            return None

        title = str(row.get("title", ""))
        discovered_str = str(row.get("discovered_date", ""))
        patched_str = str(row.get("patched_date", ""))
        
        # Parse dates
        discovered = cls._parse_date(discovered_str)
        patched = cls._parse_date(patched_str)

        # Parse severity
        sev_str = str(row.get("severity", "INFO")).strip()
        severity = SeverityLevel.from_string(sev_str)

        # Parse affected versions
        ver_str = str(row.get("affected_versions", "")).strip()
        affected: list[VersionRange] = []
        
        if ver_str:
            for part in ver_str.split(","):
                part = part.strip()
                if "-" in part and not part.endswith("+"):
                    # Range like "1.0-2.5"
                    parts = part.split("-", 1)
                    affected.append(VersionRange(min_version=parts[0], 
                                                max_version=parts[1]))
                elif "+" in part:
                    # Open-ended like "3.0+"
                    prefix, suffix = part.split("+", 1)
                    affected.append(VersionRange(min_version=prefix.strip(),
                                                max_version=suffix.strip()))
                else:
                    # Single version or range without explicit bounds
                    if "-" in part:
                        parts = part.split("-", 1)
                        affected.append(VersionRange(min_version=parts[0],
                                                    max_version=parts[1]))
                    else:
                        affected.append(VersionRange(min_version=part))

        return Vulnerability(
            cve_id=cve_id,
            title=title,
            discovered_date=discovered,
            patched_date=patched,
            affected_versions=affected,
            severity=severity,
            status=row.get("status", "open").strip() or "open"
        )

    @staticmethod
    def _parse_date(s: str) -> date | None:
        """Parse a date string into a date object."""
        if not s:
            return None
        
        # Try YYYY-MM-DD format
        try:
            return datetime.strptime(s, "%Y-%m-%d").date()
        except ValueError:
            pass

        # Try epoch timestamp
        try:
            ts = int(float(s))
            dt = datetime.fromtimestamp(ts)
            return dt.date()
        except (ValueError, OSError):
            pass

        # Try various other formats
        for fmt in ("%Y/%m/%d", "%d-%m-%Y", "%m/%d/%Y"):
            try:
                return datetime.strptime(s, fmt).date()
            except ValueError:
                continue

        # Return None if all parsing fails
        return None

    def add_vulnerabilities(self, vulns: list[Vulnerability]) -> None:
        """Add vulnerabilities to the timeline."""
        self.vulnerabilities.extend(vulns)

    def build_entries(self) -> list[TimelineEntry]:
        """Build timeline entries from all vulnerabilities.
        
        Creates one entry per significant event (discovery, patch).
        Sorts chronologically by date then by severity.
        """
        self.entries = []

        for vuln in sorted(self.vulnerabilities, key=lambda v: 
                          (v.discovered_date or date.max, -v.severity_score)):
            
            # Discovery event
            if vuln.discovered_date:
                self.entries.append(TimelineEntry(
                    date=vuln.discovered_date,
                    event_type="discovered",
                    cve_id=vuln.cve_id,
                    title=vuln.title[:100],  # Truncate for display
                    severity=vuln.severity
                ))

            # Patch event
            if vuln.patched_date:
                self.entries.append(TimelineEntry(
                    date=vuln.patched_date,
                    event_type="patched",
                    cve_id=vuln.cve_id,
                    title=f"{vuln.title[:80]}... (patched)",
                    severity=vuln.severity
                ))

            # Retired event if applicable
            if vuln.status == "retired":
                self.entries.append(TimelineEntry(
                    date=patched or vuln.discovered_date,
                    event_type="retired",
                    cve_id=vuln.cve_id,
                    title=f"{vuln.title[:80]}... (retired)",
                    severity=vuln.severity
                ))

        # Sort by date
        self.entries.sort(key=lambda e: (e.date, 
                                         {"discovered": 0, "patched": 1, "retired": 2}.get(e.event_type, 3))

        return self.entries

    def get_summary_stats(self) -> dict[str, Any]:
        """Calculate summary statistics for the timeline."""
        
        total = len(self.vulnerabilities)
        with_patch = sum(1 for v in self.vulnerabilities if v.is_patch_available)
        open_count = sum(1 for v in self.vulnerabilities if v.status == "open")

        # Patch window metrics
        patch_windows: list[int] = []
        for vuln in self.vulnerabilities:
            pw = vuln.patch_window_days
            if pw is not None and 0 < pw <= 365:  # Reasonable range
                patch_windows.append(pw)

        avg_patch_window = sum(patch_windows) / len(patch_windows) if patch_windows else 0
        max_patch_window = max(patch_windows) if patch_windows else 0
        min_patch_window = min(patch_windows) if patch_windows else 0

        # Severity distribution
        severity_dist: dict[str, int] = defaultdict(int)
        for vuln in self.vulnerabilities:
            severity_dist[vuln.severity.name] += 1

        return {
            "total_vulnerabilities": total,
            "with_patch_available": with_patch,
            "open_count": open_count,
            "patched_count": len(self.vulnerabilities) - open_count,
            "avg_patch_window_days": round(avg_patch_window, 1),
            "max_patch_window_days": max_patch_window,
            "min_patch_window_days": min_patch_window,
            "severity_distribution": dict(severity_dist),
        }

    def render_text_report(self) -> str:
        """Render a text-based report of the timeline."""
        
        lines = []
        stats = self.get_summary_stats()

        # Header
        lines.append("=" * 70)
        lines.append("VULNERABILITY DISCLOSURE TIMELINE REPORT")
        lines.append("=" * 70)
        lines.append("")

        # Summary statistics
        lines.append("SUMMARY STATISTICS")
        lines.append("-" * 40)
        
        for key, value in stats.items():
            if isinstance(value, dict):
                lines.append(f"  {key}:")
                for k, v in value.items():
                    lines.append(f"    - {k}: {v}")
            else:
                lines.append(f"  {key}: {value}")

        lines.append("")

        # Timeline entries
        if self.entries:
            lines.append("TIMELINE ENTRIES")
            lines.append("-" * 40)
            
            current_date = None
            grouped_entries: list[list[TimelineEntry]] = []
            
            for entry in self.entries:
                if entry.date != current_date:
                    if current_date is not None and grouped_entries:
                        grouped_entries.append(grouped_entries[-1])
                    
                    current_date = entry.date
                    grouped_entries.append([entry])
                else:
                    grouped_entries[-1].append(entry)

            # Format each day's entries
            for i, day_entries in enumerate(grouped_entries):
                if not day_entries:
                    continue
                
                date_str = day_entries[0].date.strftime("%Y-%m-%d")
                
                # Group by event type within the day
                by_type: dict[str, list] = defaultdict(list)
                for e in day_entries:
                    by_type[e.event_type].append(e)

                lines.append(f"\n  {date_str}")
                
                for event_type in ["discovered", "patched", "retired"]:
                    if event_type not in by_type:
                        continue
                    
                    entries = sorted(by_type[event_type], 
                                   key=lambda e: -e.severity_score)
                    
                    for entry in entries:
                        severity_marker = {
                            SeverityLevel.CRITICAL: "[CRIT]",
                            SeverityLevel.HIGH: "[HIGH]",
                            SeverityLevel.MEDIUM: "[MED]",
                            SeverityLevel.LOW: "[LOW]",
                            SeverityLevel.INFO: "[INF]",
                        }.get(entry.severity, "")

                        lines.append(f"    {severity_marker} {event_type}: " + 
                                     f"{entry.cve_id or 'N/A'} - {escape(entry.title)}")

        else:
            lines.append("No timeline entries found.")

        # Patch window analysis
        if patch_windows:
            lines.append("")
            lines.append("PATCH WINDOW ANALYSIS")
            lines.append("-" * 40)
            
            critical_vulns = [v for v in self.vulnerabilities 
                            if v.severity == SeverityLevel.CRITICAL]
            high_vulns = [v for v in self.vulnerabilities 
                         if v.severity == SeverityLevel.HIGH]

            lines.append(f"  Critical vulnerabilities: {len(critical_vulns)}")
            lines.append(f"  High vulnerabilities: {len(high_vulns)}")
            
            # Highlight slow patches
            slow_patches = [v for v in self.vulnerabilities 
                          if v.patch_window_days and v.patch_window_days > 90]
            if slow_patches:
                lines.append("")
                lines.append("  Slow patches (>90 days):")
                for v in sorted(slow_patches, key=lambda x: -x.patch_window_days)[:10]:
                    lines.append(f"    {v.cve_id}: {v.patch_window_days} days " + 
                                 f"(patched: {v.patched_date})")

        # Footer
        lines.append("")
        lines.append("=" * 70)
        lines.append("END OF REPORT")
        lines.append("=" * 70)

        return "\n".join(lines)

    def render_json_report(self, indent: int = 2) -> str:
        """Render a JSON report of the timeline."""
        
        stats = self.get_summary_stats()
        
        # Build entries as JSON-serializable objects
        entries_data = []
        for entry in self.entries:
            entries_data.append({
                "date": entry.date.isoformat(),
                "event_type": entry.event_type,
                "cve_id": entry.cve_id,
                "title": escape(entry.title),
                "severity": entry.severity.name,
            })

        report = {
            "metadata": {
                "generated_at": datetime.now().isoformat(),
                "total_vulnerabilities": len(self.vulnerabilities),
                "total_events": len(entries_data),
                **stats,
            },
            "entries": entries_data,
        }

        return json.dumps(report, indent=indent)

    def render_html_report(self) -> str:
        """Render an HTML report with basic styling."""
        
        stats = self.get_summary_stats()
        
        # Build timeline rows
        rows = []
        for entry in self.entries:
            row_class = f"event-{entry.event_type}"
            
            # Determine color based on severity and event type
            base_colors = {
                "discovered": "#e3f2fd",  # Light blue background
                "patched": "#e8f5e9",     # Light green
                "retired": "#fff3e0",     # Light orange
            }
            
            severity_colors = {
                SeverityLevel.CRITICAL: "#ffcdd2",    # Red-tinted
                SeverityLevel.HIGH: "#ffe0b2",        # Orange-tinted  
                SeverityLevel.MEDIUM: "#fff9c4",      # Yellow-tinted
                SeverityLevel.LOW: "#e1f5fe",         # Cyan-tinted
                SeverityLevel.INFO: "#